butte-build/src/ir/types.rs

```diff
@@ -0,0 +1,518 @@
+#[derive(Debug, Clone, PartialEq, TypedBuilder)]
+pub struct Rpc<'a> {
+    pub ident: DottedIdent<'a>,
+    pub methods: Vec<RpcMethod<'a>>,
+
+    #[builder(default)]
+    pub doc: ast::Comment<'a>,
+}
+
+/// A method in an RPC service.
+#[derive(Debug, Clone, PartialEq, TypedBuilder)]
+pub struct RpcMethod<'a> {
+    /// The name of the method.
+    pub ident: Ident<'a>,
+
+    /// The snake name of the method.
+    pub snake_ident: Ident<'a>,
+
+    /// The request type of the method.
+    pub request_type: DottedIdent<'a>,
+
+    /// The response type of the method.
+    pub response_type: DottedIdent<'a>,
+
+    /// Method metadata.
+    #[builder(default)]
+    pub metadata: RpcMethodMetadata,
+
+    #[builder(default)]
+    pub doc: ast::Comment<'a>,
+}
+
+#[derive(Clone, Copy, Debug, Default, PartialEq, TypedBuilder)]
+pub struct RpcMethodMetadata {
+    pub streaming: RpcStreaming,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum RpcStreaming {
+    None,
+    Client,
+    Server,
+    Bidi,
+}
+
+impl Default for RpcStreaming {
+    fn default() -> Self {
+        RpcStreaming::None
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct RootType<'a> {
+    _phantom: &'a (),
+}
+
+/// Type for `Enum` values.
+#[derive(Debug, Clone, PartialEq, Hash, Eq, From, TypedBuilder)]
+pub struct EnumVal<'a> {
+    /// The name of the enum value.
+    pub ident: Ident<'a>,
+
+    /// An optional enum value.
+    #[builder(default)]
+    pub value: Option<ast::IntegerConstant>,
+}
+
+/// Type for `Union` variants.
+#[derive(Debug, Clone, PartialEq, From, TypedBuilder)]
+pub struct UnionVariant<'a> {
+    /// The type of the variant
+    pub ty: Type<'a>,
+    /// The ident of the variant, matching the companion enum's
+    pub ident: Ident<'a>,
+}
+
+/// An identifier
+///
+/// Can be either taken from a schema, or generated synthetically,
+/// in which case an owned version of the ident is used.
+#[derive(Debug, Clone, PartialEq, Hash, Eq, AsRef, From)]
+pub struct Ident<'a> {
+    pub raw: Cow<'a, str>,
+}
+
+impl<'a> From<ast::Ident<'a>> for Ident<'a> {
+    fn from(ident: ast::Ident<'a>) -> Self {
+        Self {
+            raw: Cow::Borrowed(ident.raw),
+        }
+    }
+}
+
+impl<'a> From<&ast::Ident<'a>> for Ident<'a> {
+    fn from(ident: &ast::Ident<'a>) -> Self {
+        Self {
+            raw: Cow::Borrowed(ident.raw),
+        }
+    }
+}
+impl<'a> From<&'a str> for Ident<'a> {
+    fn from(s: &'a str) -> Self {
+        Self {
+            raw: Cow::Borrowed(s),
+        }
+    }
+}
+
+impl<'a> From<String> for Ident<'a> {
+    fn from(s: String) -> Self {
+        Self { raw: Cow::Owned(s) }
+    }
+}
+
+impl<'a> Display for Ident<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.raw.as_ref())
+    }
+}
+
+/// An identifier composed of `Ident`s separated by dots.
+#[derive(Default, Debug, Clone, PartialEq, Hash, Eq, From)]
+pub struct DottedIdent<'a> {
```

Comment:
I never liked `DottedIdent`. I think we should find a better name. I previously suggested `QualifiedIdent`. What do you think?

Reply:
Agreed, I don't like `DottedIdent` either, and `QualifiedIdent` works for me.  The method that convert simple idents to namespaced ones is called `make_fully_qualified_ident` which would also match this naming better ("fully" referring here to the fact that the namespace is specified).

Reply:
Nice, let’s make the change then, unless @cpcloud strongly opposes. I don’t remember his thoughts on this when we last discussed it.

Reply:
Gut it!

Reply:
Done!